jena-arq/src/main/java/org/apache/jena/query/ParameterizedSparqlString.java

```diff
@@ -1734,4 +1739,237 @@ public String toString() {
+            this.isParenthesisNeeded = isParenthesisNeeded;
+            this.isGrouped = false;
+        }
+
+        public ValueReplacement(String varName, Collection<List<? extends RDFNode>> groupedItems) {
+            this.varName = varName;
+            this.items = new ArrayList<>();
+            this.groupedItems = groupedItems;
+            this.isParenthesisNeeded = true;
+            this.isGrouped = true;
+        }
+
+        public String apply(String command) {
+
+            if (items.isEmpty() && groupedItems.isEmpty()) {
+                return command;
+            }
+
+            String target = createTarget(varName);
+
+            StringBuilder replacement;
+
+            if (isGrouped) {
+                replacement = groupedApply();
+            } else {
+                replacement = ungroupedApply();
+            }
+
+            return command.replace(target, replacement);
+        }
+
+        private StringBuilder groupedApply() {
+            StringBuilder replacement = new StringBuilder("");
+
+            for (List<? extends RDFNode> group : groupedItems) {
+                replacement.append("(");
+
+                for (RDFNode item : group) {
+                    String insert = createInsert(item);
+                    replacement.append(insert);
+                    replacement.append(" ");
+                }
+
+                replacement.deleteCharAt(replacement.length() - 1);
+                replacement.append(") ");
+            }
+
+            replacement.deleteCharAt(replacement.length() - 1);
+            return replacement;
+        }
+
+        private StringBuilder ungroupedApply() {
+            StringBuilder replacement = new StringBuilder("");
+
+            for (RDFNode item : items) {
+                if (isParenthesisNeeded) {
+                    replacement.append("(");
+                }
+                String insert = createInsert(item);
+                replacement.append(insert);
+                if (isParenthesisNeeded) {
+                    replacement.append(")");
+                }
+                replacement.append(" ");
+            }
+
+            replacement.deleteCharAt(replacement.length() - 1);
+
+            return replacement;
+        }
+
+        /**
+         * Tidy up varName if doesn't start with a ? or $.
+         *
+         * @param varName
+         * @return
+         */
+        private String createTarget(String varName) {
+            String target;
+
+            if (varName.startsWith("?") || varName.startsWith("$")) {
+                target = varName;
+            } else {
+                target = "?" + varName;
+            }
+            return target;
+        }
+
+        /**
+         * Insert the SPARQL representation of the RDF node.
+         *
+         * @param item
+         * @return
+         */
+        private String createInsert(RDFNode item) {
+            String insert;
+            if (item.isLiteral()) {
```

Comment:
Literals with language tags need treating specially. Lang tag literals have datatype rdf:langString, which should not be printed, and the @lang must be printed. 

Literal lexical forms need escaping for any ", newline and tabs in the string.

Other forms could be made to be more natural : strings without "^^xsd:string", integers as `123` etc.

The good news is that there is code to help at the node level so the literal and URI cases can be replaced by `FmtUtils.stringForNode(rdfNode.asNode(), (PrefixMapping)null)` - must use the "(PrefixMapping)null" form otherwise a helpful (for display) prefix mapping is used. This is the code used by the SPARQLserializer.

This includes the delimiters e.g `<...>` or string `"..."`.

The URI may need care in case "`...> finish URI early + something to follow`" is injected.